bitcoin/tests/test_rpc.py

```diff
@@ -13,24 +13,175 @@
 
 import unittest
 
-from bitcoin.rpc import Proxy
+import bitcoin.rpc
+from bitcoin.core import CBlock, CBlockHeader, lx, b2lx, COutPoint
+from bitcoin.core.script import CScript
+from bitcoin.wallet import CBitcoinAddress, CBitcoinSecret
+
+def is_active():
+    """
+    Proxy raises ValueError if cookie file not found
+    #FIXME is there a better way of doing this?
+    """
+    try:
+        p = bitcoin.rpc.Proxy()
+        p.help()
+        return True
+    except ValueError: 
+        return False
 
 class Test_RPC(unittest.TestCase):
+    _IS_ACTIVE = is_active()
     # Tests disabled, see discussion below.
     # "Looks like your unit tests won't work if Bitcoin Core isn't running;
     # maybe they in turn need to check that and disable the test if core isn't available?"
     # https://github.com/petertodd/python-bitcoinlib/pull/10
-    pass
-
-#    def test_can_validate(self):
-#        working_address = '1CB2fxLGAZEzgaY4pjr4ndeDWJiz3D3AT7'
-#        p = Proxy()
-#        r = p.validateAddress(working_address)
-#        self.assertEqual(r['address'], working_address)
-#        self.assertEqual(r['isvalid'], True)
-#
-#    def test_cannot_validate(self):
-#        non_working_address = 'LTatMHrYyHcxhxrY27AqFN53bT4TauR86h'
-#        p = Proxy()
-#        r = p.validateAddress(non_working_address)
-#        self.assertEqual(r['isvalid'], False)
+    # Sachin Meier: "I've changed it so each test checks against the _IS_ACTIVE variable "
+    #pass
+
+    def test_getbestblockhash_and_header(self):
+        if self._IS_ACTIVE:
+            proxy = bitcoin.rpc.Proxy()
+            blockhash = proxy.getbestblockhash()
+            header = proxy.getblockheader(blockhash)
+            self.assertTrue(isinstance(header, CBlockHeader))
+        else:
+            pass
+
+    def test_getblock(self):
+        if self._IS_ACTIVE:
+            proxy = bitcoin.rpc.Proxy()
+            blockhash = proxy.getbestblockhash()
+            # Test from bytes
+            block1 = proxy.getblock(blockhash)
+            self.assertTrue(isinstance(block1, CBlock))
+            # Test from str
+            block2 = proxy.getblock("0000000000000000000b4b0daf89eac9d84138fc900b8c473d4da70742e93dd0")
+            self.assertTrue(isinstance(block2, CBlock))
+        else:
+            pass
+
+    def test_getblockcount_et_al(self):
+        # This test could possibly false-fail if new blocks arrive. 
+        # Highly unlikely since they're quick calls
```

Comment:
I don't think "highly unlikely" is a good argument. It will someday fail and confuse someone. Tests should be deterministic.

By the way, because the tests should be deterministic, I don't think that just connecting to with the default parameters is a good idea - the tests will depend on environment that they run in, and what `bitcoin.conf` is there is unpredictable. I would introduce a special environment variable with unique prefix, like `BITCOINLIB_TEST_RUN_ENABLE_RPC_CONF=...some_test_bitcoin.conf` or something like that, that would indicate that the environment is ready for the tests, and should use a special config. This way an automatic test service like travis ci can be used to setup a test environment where RPC can be tested, but the tests won't have any unexpected effects in other environments.

In general, user would not expect that some tests could even connect to their working bitcoind, for example, and may not be happy about that even if the commands are innoulous. On the other hand, the connection might block if the daemon port is filtered via dropping the packets rather than sending RST back, and then the test will be stuck for quite a long time.

Reply:
Sorry, this is confusing for me. Are you saying we should add a separate ```bitcoin.conf``` just for when a user runs rpc tests? I'm not sure there's a failsafe way to test ```getblockcount```. Do you have any ideas?

Reply:
I propose to use a specially-named environment variable to point to a separate bitcoin.conf that would contain the parameters for the test. If an environment variable is not specified, the RPC tests just not run. If it is specified, a failed connection means error.

You can safely test blockcount on a regtest node. As I said in another comment, relying on a mainnet node will be unrealistic if you want your tests to be run in some automatic test system like Travis CI - your test container would need to include a full node into it. Even if you manage that, a full node that connects to the network is a dynamic thing that won't be cached and your test times will be very long

Reply:
how do you set the chain with a cli command? Is this possible?

Reply:
What do you think about just removing this test? ```getblockcount``` seems easy enough to get right.

Reply:
> I propose to use a specially-named environment variable to point to a separate bitcoin.conf that would contain the parameters for the test. If an environment variable is not specified, the RPC tests just not run. If it is specified, a failed connection means error.

bitcoind's `test_framework` has some functionality for this out of the box. Should probably be packaged into a python library at some point.

Reply:
The idea is that you would prepare your test environment in a way that a regtest bitcoind is already running and a special bitcoin.conf already points on its rpc. you indicate that with a special env var. You won't need to set the chain with a command. This is workable when you have automatic testing in a container. When manually running the tests, you will usually just not set that special env var and will skip the rpc tests.

> What do you think about just removing this test? 

I have no opinion - I'm not sure that these tests are needed at all, and I removed that functionality from bitcointx entirely in favour of a dumb RPC proxy without converting any values. Others might hold different approach to this, and that's OK. But I saw a few potential problems with the code, that I thought are worth highlighting.

Reply:
Okay. I am not very experienced with writing tests, so I'm getting a bit lost here. @kanzure, by your comment, did you mean there's a tool I can use to facilitate this?

Reply:
https://github.com/bitcoin/bitcoin/tree/master/test/functional/test_framework

You don't have to use this; porting it over to python-bitcoinlib would be a bit of a project. But I wanted to point out that it exists and helps with testing (setting up multiple nodes, wiring them up, etc).